### PR TITLE
Add initial draft of ideas page.

### PR DESCRIPTION
**Pull request by hameerabbasi**

### CODE REVIEW

File: docs/gsoc2020_ideas.rst [hunk 1]
```diff
@@ -0,0 +1,134 @@
...
+
+If you have an idea of what you would like to work on (see below for ideas)
+and are considering participating:
+
+
+1. Read the `PSF page <http://python-gsoc.org/>`_ carefully, it contains
+   important advice on the process.
+2. Read `advice on writing a proposal <http://turnbull.sk.tsukuba.ac.jp/Blog/SPAM.txt#how-to-spam-in-detail>`_
+   (written with the Mailman project in mind, but generally applicable)
+3. Look at the guidelines on how to contribute to Scipy.
```
Comment (rgommers) [inline docs/gsoc2020_ideas.rst]: this isn't applicable

File: docs/gsoc2020_ideas.rst [hunk 2]
```diff
@@ -0,0 +1,134 @@
...
+4. Make a enhancement/bugfix/documentation fix -- it does not have to be big,
+   and it does not need to be related to your proposal. Doing so before
+   applying for the GSoC is a hard requirement for ``uarray``. It helps
+   everyone you get some idea how things would work during GSoC.
+5. Start writing your proposal early, post a draft to the the mentors' email
+   addresses mailing list and iterate based on the feedback you receive. This
```
Comment (rgommers) [inline docs/gsoc2020_ideas.rst]: "mentors's addresses mailing list" needs a rephrase

File: docs/gsoc2020_ideas.rst [hunk 3]
```diff
@@ -0,0 +1,134 @@
...
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+This idea would allow a backend to encompass functions from more than one domain.
+
+The primary goal of this project would be:
+
+* Develop a system that allows, via some kind of matching mechanism, to select
+  which domains it supports, **while maintaining backward compatibility**.
+
+This would allow a backend targeting NumPy to also target, for example, the
+``numpy.random`` submodule.
```
Comment (rgommers) [inline docs/gsoc2020_ideas.rst]: I'm not sure this clarifies it. If you're targeting (all of) NumPy, you're already targeting `numpy.random` as well?

This whole idea needs a bit more detail, it's unclear to me from a single read what the work would entail.

File: docs/gsoc2020_ideas.rst [hunk 4]
```diff
@@ -0,0 +1,134 @@
...
+
+This project has a somewhat complicated design and needs some involved
+implementation work, and will require interacting with the mentors to flesh
+out and work through.
+
+* Required knowledge: Python C-API and C++
+* Difficulty level: easy
```
Comment (rgommers) [inline docs/gsoc2020_ideas.rst]: doesn't sound easy - medium at least?

File: docs/gsoc2020_ideas.rst [hunk 5]
```diff
@@ -0,0 +1,134 @@
...
+
+* Allow a downstream library to detect a backend and run specialised code for
+  it.
+* Allow a downstream library to fail-fast on a known-unsupported backend.
+
+This project has a straightforward design and needs some implementation work,
+and will require interacting with the mentors to implement and polish.
+
+* Required knowledge: Python C-API and C++
+* Difficulty level: easy
```
Comment (rgommers) [inline docs/gsoc2020_ideas.rst]: probably also medium difficulty at least

File: docs/gsoc2020_ideas.rst [hunk 6]
```diff
@@ -0,0 +1,134 @@
...
+* Potential mentors: Peter Bell and Hameer Abbasi
+
+``unumpy``: Expand overall coverage
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+This project is split into two parts:
+
+* Adding further converage of the NumPy API.
```
Comment (rgommers) [inline docs/gsoc2020_ideas.rst]: typo in "coverage"

File: docs/gsoc2020_ideas.rst [hunk 7]
```diff
@@ -0,0 +1,134 @@
...
+* Adding more backends to NumPy.
```
Comment (rgommers) [inline docs/gsoc2020_ideas.rst]: `NumPy` -> `unumpy` right?

File: docs/gsoc2020_ideas.rst [hunk 8]
```diff
@@ -0,0 +1,134 @@
...
+
+We realise this is a large (possibly open-ended) undertaking, and so there
+will need to be a minimum amount of work done in order to pass.
```
Comment (rgommers) [inline docs/gsoc2020_ideas.rst]: This isn't a statement that a student can do much with. I suggest to instead say something like "~150 functions left to implement, if time left we would like to see a new backend for Ray implemented".

File: docs/gsoc2020_ideas.rst [hunk 9]
```diff
@@ -0,0 +1,134 @@
...
+3. Adding tests.
+4. Adding documentation on use.
+5. Publishing a final version to PyPI.
+
+This project has a somewhat some minimal design and needs some involved
+implementation work, and will require interacting with the mentors to flesh
+out and work through.
+
+* Required knowledge: Python (intermediate level) and calculus
+* Difficulty level: easy
```
Comment (rgommers) [inline docs/gsoc2020_ideas.rst]: medium?

File: docs/gsoc2020_ideas.rst (hunk 4 above)
Comment (rgommers) [reply docs/gsoc2020_ideas.rst]: after looking at the rest: this is probably the hardest one. perhaps "hard"

**Comment by rgommers** [comment]
Also, you may want to put this in a wiki page rather than merging it into the docs, given that it's time-sensitive content; wiki can remain as is, docs you'll want to clean again later.

File: docs/gsoc2020_ideas.rst (hunk 3 above)
Comment (hameerabbasi) [reply docs/gsoc2020_ideas.rst]: Is the text I added enough?

**Comment by hameerabbasi** [comment]
> Also, you may want to put this in a wiki page rather than merging it into the docs, given that it's time-sensitive content; wiki can remain as is, docs you'll want to clean again later.

I gave the path a bit more structure, so it won't change much.

File: docs/gsoc2020_ideas.rst (hunk 3 above)
Comment (rgommers) [reply docs/gsoc2020_ideas.rst]: Yes, clear now - thanks

**Pull request merged**